src/facilities-context/validators.tsx

```diff
@@ -0,0 +1,102 @@
+import { differenceInCalendarDays } from "date-fns";
+import { maxBy } from "lodash";
+
+import { validateCumulativeCases } from "../infection-model/validators";
+import { ModelInputs } from "../page-multi-facility/types";
+
+type SingleDayValidator = (
+  modelInputs: ModelInputs | null,
+) => ModelInputs | null;
+
+type HistoryValidator = (modelVersions: ModelInputs[]) => ModelInputs[];
+
+function notNull(version: ModelInputs | null): version is ModelInputs {
+  return version !== null;
+}
+
+const removeNonsenseCases: SingleDayValidator = (modelInputs) => {
+  if (modelInputs?.isReference) {
+    if (modelInputs.ageUnknownCases) {
+      if (modelInputs.ageUnknownCases < 0) return null;
+      if (
+        modelInputs.ageUnknownPopulation !== undefined &&
+        modelInputs.ageUnknownCases > modelInputs.ageUnknownPopulation
+      )
+        return null;
+    }
+  }
+  return modelInputs;
+};
+
+const getPrecedingVersion = (
+  currentObservedAt: Date,
+  modelInputVersions: ModelInputs[],
+) => {
+  return maxBy(
+    modelInputVersions.filter(
+      (version) =>
+        differenceInCalendarDays(currentObservedAt, version.observedAt) >= 1,
+    ),
+    (version) => version.observedAt,
+  );
+};
+
+const isMonotonicIncrease = (
+  modelInputs: ModelInputs,
+  modelInputVersions: ModelInputs[],
+): boolean => {
+  const currentObservedAt = modelInputs.observedAt;
+  const previous = getPrecedingVersion(currentObservedAt, modelInputVersions);
+
+  return validateCumulativeCases(modelInputs, {
+    previous,
+    compareSums: modelInputs.isReference || previous?.isReference,
+  });
+};
+
+const ensureCasesIncreaseMonotonically: HistoryValidator = (modelVersions) => {
+  const validModelVersions = [...modelVersions];
+  for (const version of modelVersions) {
+    if (!isMonotonicIncrease(version, validModelVersions)) {
+      if (version.isReference) {
+        validModelVersions.splice(validModelVersions.indexOf(version), 1);
+      } else {
+        // drop one or more preceding reference versions instead
+        // until monotonicity is achieved (or we hit another user version)
+        let monotonic = false;
+        while (!monotonic) {
+          const prev = getPrecedingVersion(
+            version.observedAt,
+            validModelVersions,
+          );
+          if (!prev || !prev.isReference) {
+            break;
```

Comment:
no, in this case we do want to `break` ... `!prev` means there are no previous cases, so if we `continue` we will enter an infinite loop. and if we hit another piece of user data we let it stand regardless of validity, because we want all user data to be visible. I can comment this logic better if you think it's still too obscure

Reply:
I get it now.  Thanks for the explanation.  No additional comment necessary